src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -102,6 +89,128 @@ pub struct ConfigSpace {
 // SAFETY: `ConfigSpace` contains only PODs in `repr(C)` or `repr(transparent)`, without padding.
 unsafe impl ByteValued for ConfigSpace {}
 
+#[derive(Debug, thiserror::Error, displaydoc::Display)]
+enum AddRxBufferError {
+    /// Error while parsing new buffer: {0}
+    Parsing(#[from] IoVecError),
+    /// RX buffer is too small
+    BufferTooSmall,
+}
+
+/// A map of all the memory the guest has provided us with for performing RX
+#[derive(Debug)]
+pub(crate) struct RxBuffers {
+    // minimum size of a usable buffer for doing RX
+    min_buffer_size: u32,
+    // An [`IoVecBufferMut`] covering all the memory we have available for receiving network
+    // frames.
+    iovec: IoVecBufferMut,
+    // A map of which part of the memory belongs to which `DescriptorChain` object
+    parsed_descriptors: VecDeque<ParsedDescriptorChain>,
+    // Buffers that we have used and they are ready to be given back to the guest.
+    deferred_descriptor: Option<ParsedDescriptorChain>,
```

Comment:
I take a bit of a philosophical issue with this field being used for non-deferred descriptors (e.g. we assign each descriptor we use to this field temporarily, but they only become "deferred" if we hit the ratelimiter in a check that happens after the assignment to this variable). 

I take a second issue with the `rx_bytes_read` variable in `Net`. That variable I think should move into `RxBuffers`, because really, it now means `rx_bytes_deferred`. BUT, I don't think we need this `rx_bytes_*` variable at all, because it has the same value as the `length` of the deferred descriptor. Additionally, if we only set `deferred_descriptor` while we actually are deferring a descriptor, then the `rx_deferred_descriptor` bool in `Net` also becomes useless, as it will just be `rx_buffers.deferred_descriptor.is_some()`.

So my proposal is the following: Change `.add_used` and `.read_from_mmds_or_tap` to returned the `ParsedDescriptor` that was written to. Change `rate_limited_rx_single_frame` and `finish_frame` to take a `ParsedDescriptor` parameter. If the ratelimits are insufficient, then set `deferred_descriptor` only in this function.

Essentially, the following diff implements this idea (compile-tested only, and unittests blissfully ignored):

<details>

```diff
diff --git a/src/vmm/src/devices/virtio/net/device.rs b/src/vmm/src/devices/virtio/net/device.rs
index f42c34c0d..adeec3422 100755
--- a/src/vmm/src/devices/virtio/net/device.rs
+++ b/src/vmm/src/devices/virtio/net/device.rs
@@ -152,7 +152,7 @@ impl RxBuffers {
     /// The `size` should be the number of bytes previously written in the [`RxBuffers`]. This
     /// means that `size` cannot be more than the currently available bytes in the underlying
     /// `IoVecBufferMut`.
-    fn mark_used(&mut self, size: u32, metrics: &NetDeviceMetrics) {
+    fn mark_used(&mut self, size: u32, metrics: &NetDeviceMetrics) -> ParsedDescriptorChain {
         // Since we were able to write a frame in guest memory, we should have at least one
         // descriptor chain here. If not, we have a bug, so fail fast, since the device is
         // fundamentally broken.
@@ -174,7 +174,7 @@ impl RxBuffers {
         }
         self.header_set_num_buffers(1);
         self.iovec.drop_descriptor_chain(&parsed_dc);
-        self.deferred_descriptor = Some(parsed_dc);
+        parsed_dc
     }
 
     /// Write the number of descriptors used in VirtIO header
@@ -193,20 +193,15 @@ impl RxBuffers {
 
     /// This will let the guest know that about all the `DescriptorChain` object that has been
     /// used to receive a frame from the TAP.
-    fn finish_frame(&mut self, rx_queue: &mut Queue) -> bool {
-        match self.deferred_descriptor.take() {
-            Some(used_dc) => {
-                if let Err(err) = rx_queue.add_used(used_dc.head_index, used_dc.length) {
-                    error!(
-                        "net: Failed to add used descriptor {} of length {} to RX queue: {err}",
-                        used_dc.head_index, used_dc.length
-                    );
-                    false
-                } else {
-                    true
-                }
-            }
-            None => false,
+    fn finish_frame(&mut self, rx_queue: &mut Queue, frame: &ParsedDescriptorChain) -> bool {
+        if let Err(err) = rx_queue.add_used(frame.head_index, frame.length) {
+            error!(
+                "net: Failed to add used descriptor {} of length {} to RX queue: {err}",
+                frame.head_index, frame.length
+            );
+            false
+        } else {
+            true
         }
     }
 }
@@ -231,9 +226,6 @@ pub struct Net {
     pub(crate) rx_rate_limiter: RateLimiter,
     pub(crate) tx_rate_limiter: RateLimiter,
 
-    pub(crate) rx_deferred_frame: bool,
-
-    rx_bytes_read: usize,
     rx_frame_buf: [u8; MAX_BUFFER_SIZE],
 
     tx_frame_headers: [u8; frame_hdr_len()],
@@ -299,8 +291,6 @@ impl Net {
             queue_evts,
             rx_rate_limiter,
             tx_rate_limiter,
-            rx_deferred_frame: false,
-            rx_bytes_read: 0,
             rx_frame_buf: [0u8; MAX_BUFFER_SIZE],
             tx_frame_headers: [0u8; frame_hdr_len()],
             irq_trigger: IrqTrigger::new().map_err(NetError::EventFd)?,
@@ -422,15 +412,18 @@ impl Net {
     // Attempts to copy a single frame into the guest if there is enough
     // rate limiting budget.
     // Returns true on successful frame delivery.
-    fn rate_limited_rx_single_frame(&mut self) -> bool {
+    fn rate_limited_rx_single_frame(&mut self, frame: ParsedDescriptorChain) -> bool {
         let rx_queue = &mut self.queues[RX_INDEX];
-        if !Self::rate_limiter_consume_op(&mut self.rx_rate_limiter, self.rx_bytes_read as u64) {
+        if !Self::rate_limiter_consume_op(&mut self.rx_rate_limiter, frame.length as u64) {
             self.metrics.rx_rate_limiter_throttled.inc();
+            self.rx_buffer.deferred_descriptor = Some(frame);
             return false;
         }
 
-        if !self.rx_buffer.finish_frame(rx_queue) {
-            Self::rate_limiter_replenish_op(&mut self.rx_rate_limiter, self.rx_bytes_read as u64);
+        if !self.rx_buffer.finish_frame(rx_queue, &frame) {
+            Self::rate_limiter_replenish_op(&mut self.rx_rate_limiter, frame.length as u64);
+            self.rx_buffer.deferred_descriptor = Some(frame);
+            return false;
         }
         true
     }
@@ -550,7 +543,7 @@ impl Net {
     }
 
     // We currently prioritize packets from the MMDS over regular network packets.
-    fn read_from_mmds_or_tap(&mut self) -> Result<usize, NetError> {
+    fn read_from_mmds_or_tap(&mut self) -> Result<ParsedDescriptorChain, NetError> {
         // If we don't have any buffers available try to parse more from the RX queue. There might
         // be some buffers we didn't get the chance to process, because we got to handle the TAP
         // event before the RX queue event.
@@ -560,7 +553,7 @@ impl Net {
             // If after parsing the RX queue we still don't have any buffers stop processing RX
             // frames.
             if self.rx_buffer.len() == 0 {
-                return Ok(0);
+                return Err(NetError::NoRxDescriptors)
             }
         }
 
@@ -575,33 +568,29 @@ impl Net {
                 self.rx_buffer
                     .iovec
                     .write_all_volatile_at(&self.rx_frame_buf[..vnet_hdr_len() + len], 0)?;
-                self.rx_buffer
-                    .mark_used((vnet_hdr_len() + len).try_into().unwrap(), &self.metrics);
-                return Ok(vnet_hdr_len() + len);
+                return Ok(self.rx_buffer
+                    .mark_used((vnet_hdr_len() + len).try_into().unwrap(), &self.metrics))
             }
         }
 
         // SAFETY: this is safe because we ensured that `self.rx_buffer` has at least one
         // DescriptorChain parsed in it.
         let len = unsafe { self.read_tap().map_err(NetError::IO) }?;
-        self.rx_buffer
-            .mark_used(len.try_into().unwrap(), &self.metrics);
-        Ok(len)
+        Ok(self.rx_buffer
+            .mark_used(len.try_into().unwrap(), &self.metrics))
     }
 
     /// Read as many frames as possible.
     fn process_rx(&mut self) -> Result<(), DeviceError> {
         loop {
             match self.read_from_mmds_or_tap() {
-                Ok(0) => {
+                Err(NetError::NoRxDescriptors) => {
                     self.metrics.no_rx_avail_buffer.inc();
                     break;
                 }
-                Ok(count) => {
-                    self.rx_bytes_read = count;
+                Ok(parsed_chain) => {
                     self.metrics.rx_count.inc();
-                    if !self.rate_limited_rx_single_frame() {
-                        self.rx_deferred_frame = true;
+                    if !self.rate_limited_rx_single_frame(parsed_chain) {
                         break;
                     }
                 }
@@ -627,25 +616,15 @@ impl Net {
         self.try_signal_queue(NetQueue::Rx)
     }
 
-    // Process the deferred frame first, then continue reading from tap.
-    fn handle_deferred_frame(&mut self) -> Result<(), DeviceError> {
-        if self.rate_limited_rx_single_frame() {
-            self.rx_deferred_frame = false;
-            self.rx_buffer.deferred_descriptor = None;
-            // process_rx() was interrupted possibly before consuming all
-            // packets in the tap; try continuing now.
-            return self.process_rx();
-        }
-
-        self.try_signal_queue(NetQueue::Rx)
-    }
-
     fn resume_rx(&mut self) -> Result<(), DeviceError> {
-        if self.rx_deferred_frame {
-            self.handle_deferred_frame()
-        } else {
-            self.process_rx()
+        if let Some(frame) = self.rx_buffer.deferred_descriptor.take() {
+            if !self.rate_limited_rx_single_frame(frame)
+            {
+                return self.try_signal_queue(NetQueue::Rx)
+            }
         }
+
+        self.process_rx()
     }
 
     fn process_tx(&mut self) -> Result<(), DeviceError> {
@@ -706,7 +685,7 @@ impl Net {
                 &self.metrics,
             )
             .unwrap_or(false);
-            if frame_consumed_by_mmds && !self.rx_deferred_frame {
+            if frame_consumed_by_mmds && !self.rx_buffer.deferred_descriptor.is_some() {
                 // MMDS consumed this frame/request, let's also try to process the response.
                 process_rx_for_mmds = true;
             }
@@ -836,16 +815,7 @@ impl Net {
             return;
         }
 
-        if self.rx_deferred_frame
-        // Process a deferred frame first if available. Don't read from tap again
-        // until we manage to receive this deferred frame.
-        {
-            self.handle_deferred_frame()
-                .unwrap_or_else(|err| report_net_event_fail(&self.metrics, err));
-        } else {
-            self.process_rx()
-                .unwrap_or_else(|err| report_net_event_fail(&self.metrics, err));
-        }
+        self.resume_rx().unwrap_or_else(|err| report_net_event_fail(&self.metrics, err));
     }
 
     /// Process a single TX queue event.
diff --git a/src/vmm/src/devices/virtio/net/mod.rs b/src/vmm/src/devices/virtio/net/mod.rs
index e8a3f86ac..cefccc6cf 100644
--- a/src/vmm/src/devices/virtio/net/mod.rs
+++ b/src/vmm/src/devices/virtio/net/mod.rs
@@ -58,4 +58,6 @@ pub enum NetError {
     VnetHeaderMissing,
     /// IoVecBuffer(Mut) error: {0}
     IoVecError(#[from] IoVecError),
+    /// No rx descriptors available
+    NoRxDescriptors,
 }

```

</details>

Now, admittedly, I don't know if this idea is compatible with future extentions for mergable buffers

Reply:
I remember trying to do that and having some issues with error handling and error propagation, etc., but some time has passed since, so let me try to do that again now. If nothing breaks, I'm happy to commit it.

Reply:
mrg_rxbuf will not like this as there we will have to return a `Vec<ParsedDescriptor>` and this is not the way we should go.
I have a commit in my previous PR that deals with deferred variable: https://github.com/firecracker-microvm/firecracker/pull/4813/commits/a267a0ce3ab8790f1ad397e83faf2fdbbf199b96

Reply:
> mrg_rxbuf will not like this as there we will have to return a `Vec<ParsedDescriptor>` and this is not the way we should go. I have a commit in my previous PR that deals with deferred variable: [a267a0c](https://github.com/firecracker-microvm/firecracker/commit/a267a0ce3ab8790f1ad397e83faf2fdbbf199b96)

Mh, so in https://github.com/firecracker-microvm/firecracker/commit/06d97cdd92bd40cdf8c485bc097c5e7bd8c75646 you also move the writing of the used ring entries to `mark_used` (instead of only doing it in `finish_packet`), and in that case we can just use a `(bytes_written, nr_descriptors_used)` tuple instead of a `ParsedDescriptorChain`, and the rest of my diff would look pretty much the same, I think

Reply:
ok, in any case, seems like this can be a follow up PR.

Reply:
> ok, in any case, seems like this can be a follow up PR.

I disagree, this is mostly about code introduced in this PR, and imo it makes the logic a lot easier to follow because it becomes less stateful.

Reply:
Ok, I implemented a version of this suggestion.

I renamed `deferred_frame` to `used_descriptor`. It does not exactly address you philosophical issue :) but my rationale here is that `used_descriptor` will hold that last used `DescriptorChain`, until someone calls `finish_frame`. Then, we can simply drop the `rx_deferred_descriptor` `bool` and substitute it with the `is_some()` condition you mentioned.

I did not have `mark_used` to return anything. What we need is the number of bytes written in guest memory, however that we already know at the call site; it is the `size` argument of `mark_used` itself. This way we can also, trivially drop the `rx_bytes_read` as well, i.e. now, `rate_limited_rx_single_frame` takes the same `size` argument.

I think this is trivially extended for mergeable buffers, WDYT?

Reply:
Yeah, I think resolving the duplication of the fields in Net/RxBuffers addresses my main concern, so the rest of this is a nit. I will try to motivate my proposed changes again though xP: Right now, what does it mean to call `finish_frame` if `used_descriptor` is `None`? The type system doesn't stop us from doing that, but it would be a logic bug in our code. If we instead have `finish_frame` take all the information it needs as (non-optional) parameters, then we eliminate this class of bugs at compile time.

Also, if we do a follow up, can we untangle `resume_rx`, `handle_deferred_frame` and `process_rx_events` like I was forced to do in my diff above? Right now they're quite hard to follow imo because they all contain very similar logic to check for deferred frames, while also calling each other.

Reply:
Ok, let me have another go with it :)